typescript/sdk/src/app.ts

```diff
@@ -1,44 +1,44 @@
-import { ContractsBuilder, IAbacusContracts } from './contracts';
-import { MultiProvider } from './provider';
+import {
+  AbacusAddresses,
+  AbacusContracts,
+  AbacusFactories,
+  addresses,
+  attach,
+  connect,
+} from './contracts';
 import { ChainMap, ChainName, Connection } from './types';
 import { MultiGeneric, objMap } from './utils';
 
 export class AbacusApp<
-  Contracts extends IAbacusContracts<any, any>,
+  Contracts extends AbacusContracts,
   Chain extends ChainName = ChainName,
 > extends MultiGeneric<Chain, Contracts> {
-  constructor(
-    builder: ContractsBuilder<any, Contracts>,
-    contractAddresses: ChainMap<Chain, any>,
-    readonly multiProvider: MultiProvider<Chain>,
-  ) {
-    super(
-      objMap(
-        contractAddresses,
-        (chain, addresses) =>
-          new builder(
-            addresses,
-            multiProvider.getChainConnection(chain).getConnection()!,
-          ),
-      ),
-    );
+  constructor(public contractsMap: ChainMap<Chain, Contracts>) {
+    super(contractsMap);
   }
 
-  public contractsMap = this.chainMap;
+  static buildContracts<
+    Contracts extends AbacusContracts,
+    Chain extends ChainName,
+  >(
+    addressesMap: ChainMap<Chain, AbacusAddresses>,
+    factories: AbacusFactories,
+  ): ChainMap<Chain, Contracts> {
+    return objMap(
+      addressesMap,
+      (_, addresses) => attach(addresses, factories) as Contracts,
+    );
+  }
 
-  getContracts(
-    chain: Chain,
-  ): Contracts extends IAbacusContracts<any, infer C> ? C : never {
-    return this.get(chain).contracts;
+  getContracts(chain: Chain): Contracts {
+    return this.get(chain);
   }
 
-  getAddresses(
-    chain: Chain,
-  ): Contracts extends IAbacusContracts<infer A, any> ? A : never {
-    return this.get(chain).addresses;
+  getAddresses(chain: Chain): AbacusAddresses {
+    return addresses(this.get(chain));
   }
 
   reconnect(chain: Chain, connection: Connection): void {
-    this.get(chain).reconnect(connection);
+    connect(this.get(chain), connection);
```

Comment:
Isn't the underlying connect non-mutative? Thus `reconnect` is non-mutative, but the void makes me thing it would be mutative?